### PR TITLE
require only within required scope

### PR DESCRIPTION
**Pull request by charlierudolph**
@alexdavid @kevgo

Wanted to get your take on an experiment I made where I put all requires at the top of whatever scope they were needed instead of putting them all at the top of the file.

Makes some functions a bit longer but highlights the complexity of each function.

### CODE REVIEW

File: src/cli.coffee [hunk 1]
```diff
@@ -1,11 +1,5 @@
-asyncHandlers = require 'async-handlers'
-{docopt} = require 'docopt'
-fs = require 'fs-extra'
-path = require 'path'
 extensions = require './config/supported_file_extensions'
-
-
-options = docopt """
+usage = """
```
Comment (kevgo) [inline src/cli.coffee]: This variable is used only once. You could keep the require of docopt here, and provide the usage string directly to it. I think that's still very readable.

File: src/cli.coffee [hunk 2]
```diff
@@ -15,18 +9,25 @@ options = docopt """
...
 
 
-generateConfig = ->
-  break for extension in extensions when options[extension]
+{docopt} = require 'docopt'
+options = docopt usage
+
+
+generateConfig = (extension) ->
+  asyncHandlers = require 'async-handlers'
```
Comment (kevgo) [inline src/cli.coffee]: `asyncHandlers` is used in the callback, not here. Thinking your approach all the way to the end, shouldn't it be required on line 24?

File: src/cli.coffee [hunk 3]
```diff
@@ -15,18 +9,25 @@ options = docopt """
...
+  fsExtra = require 'fs-extra'
```
Comment (kevgo) [inline src/cli.coffee:14]: Move this to line 26?

File: src/cli.coffee [hunk 4]
```diff
@@ -15,18 +9,25 @@ options = docopt """
...
     asyncHandlers.exitOnError err
     console.log "Configuration file generated at \"#{destFilename}\""
-  fs.copy src, dest, callback
+  fsExtra.copy src, dest, callback
 
 
 if options['--generate-config']
-  generateConfig()
+  break for extension in extensions when options[extension]
+  generateConfig extension
```
Comment (kevgo) [inline src/cli.coffee]: This seems an unrelated changes, can you extract it out of this PR, so that it is easier to really only see the changes from inlining requires here?

File: src/config/configuration_loader.coffee [hunk 5]
```diff
@@ -38,11 +37,16 @@ class ConfigurationLoader
 
 
   loadDefaultConfig: (done) =>
+    fsExtra = require 'fs-extra'
     fsExtra.readJson @defaultConfigPath, done
 
 
   loadUserConfig: (done) =>
-    filePaths = _.map extensions, (ext) => path.join @dir, "dependency-lint.#{ext}"
+    async = require 'async'
```
Comment (kevgo) [inline src/config/configuration_loader.coffee]: I'm seeing a good amount of repetition here. `async` is defined and imported in many different methods, as is `fs` and `fsExtra`, `path`, etc. 

This approach is similar to the question where variables should be declared (in languages that require that): right before they are used, or all together at the top of the file/method? I personally prefer the former, so in general I like the thinking here. 

Having `require`s inline has the advantage that it becomes much easier to clean out unused dependencies while refactoring code. Dependency-Lint then helps remove them from `package.json`. This is a big advantage. And it makes it easier to see what each variable that is used in code is, and where it comes from. That's big too. 

On the downside, this makes me a bit concerned that different methods use different alternative npm modules. Example: on order to display colored terminal output, method a imports and uses `cli-color`, method 2 uses `colors`, and method 3 uses `chalk`. Because they are not aware that the other method already uses a library. Arguably, if we keep file sizes at a sane level, this argument isn't very strong.

File: src/linter.coffee [hunk 6]
```diff
@@ -1,38 +1,44 @@
...
 
 
-  lint: (done) ->
-    async.parallel {
-      listedModules: @getListedModules
-      usedModules: @usedModuleFinder.find
-    }, asyncHandlers.transform(@dependencyLinter.lint, done)
+  lint: (done) =>
+    async = require 'async'
+    asyncHandlers = require 'async-handlers'
```
Comment (kevgo) [inline src/linter.coffee]: Another disadvantage of inlining the requires here is that it artificially increases the cyclomatic complexity score for methods, which will lead to the linter yelling at us more for no good reason. 

Another disadvantage is that it makes methods longer and harder to read. 

And it makes using external libraries more cumbersome. If I have to require `async` each time I want to run one of the many async methods, that can quickly add a lot of boilerplate to a code base (like 10-20% of total lines, if we use a lot of external libraries, as we should). I can see that this can make me not want to use async or other libraries that much anymore, which would be a sad outcome.

File: src/linter.coffee [hunk 7]
```diff
@@ -1,38 +1,44 @@
...
+    async.auto {
+      packageJson: @getPackageJson
+      usedModules: @getUsedModules
+    }, asyncHandlers.transform(@lintModules, done)
 
 
   extractListedModules: (packageJson) ->
+    _ = require 'lodash'
```
Comment (kevgo) [inline src/linter.coffee:19]: Another observation: This method now does more than one thing. Before, it was simply massaging some data. Now it imports a library, and then uses that library to massage some data. Following refactoring rules, I'd say this should be extracted.

File: src/linter.coffee [hunk 8]
```diff
@@ -1,38 +1,44 @@
...
+    dependencyLinter.lint {listedModules, usedModules}
+
+
+  getPackageJson: (done) =>
+    fsExtra = require 'fs-extra'
+    path = require 'path'
     filePath = path.join @dir, 'package.json'
-    callback = asyncHandlers.transform @extractListedModules, done
-    fsExtra.readJson filePath, callback
+    fsExtra.readJson filePath, done
```
Comment (kevgo) [inline src/linter.coffee:40]: This method could be a nice single line in the old way. Now its 4 lines, 3 of them are boilerplate and ceremony.

File: src/linter/used_module_finder/index.coffee [hunk 9]
```diff
@@ -1,27 +1,35 @@
...
+    asyncHandlers = require 'async-handlers'
+    handler = asyncHandlers.transform @normalizeModules, done
+    async.parallel {
+      executedModules: @getExecutedModules
+      requiredModules: @getRequiredModules
+    }, handler
+
+
+  # Private
+  getExecutedModules: (done) =>
```
Comment (kevgo) [inline src/linter/used_module_finder/index.coffee:17]: Using a leading underscore for private methods has the advantage that it becomes clear if a private method is accessed from the outside. It also makes clear who is private and who is not when debugging/inspecting objects. I think overall we had pretty good experiences with this pattern on ASF. How do you feel about adopting this pattern here?

**Comment by kevgo** [comment]
Cool, interesting idea @charlierudolph, thanks for doing the research! I have to say I was initially in favor of it, since it matches the way I declare variables inline. Seeing this applied on a well structured and well written code base that makes good use of third-party libraries, I have to say I don't like the idea as much anymore. It introduces too much boilerplate into too many methods. I estimate around 20-30% in the average (considering that methods are < 10 loc, often a lot less, and each method uses between 1-3 external libraries. This is especially true for super versatile libraries like lodash and async. 

I also don't like that it makes methods more complex, and makes them do several things now. This stood out to me a lot when reviewing this PR. 

The biggest benefit I see here is that it makes it easier to clean up unused dependencies. I think a good linter that points out unused variables can do the same for us. 

So, taking all of this into consideration, I'm leaning against this idea. Its a great insight though, and good learning, so I'm happy that we did this experiment! Improving code structure is a never-ending journey!

File: src/linter/used_module_finder/index.coffee (hunk 9 above)
Comment (charlierudolph) [inline src/linter/used_module_finder/index.coffee:17]: Meant to remove that comment. Lets save that for another discussion.

File: src/linter.coffee (hunk 6 above)
Comment (charlierudolph) [inline src/linter.coffee]: The first point is something I like about it. The complexity is the same to me as you are still using the same variables its just moving where something is defined.

No rebuttal for the second point that is a downside to this approach

To the third point, I'm leaning towards moving anything that is required multiple times to an outer scope so that it can stay dry

File: src/cli.coffee (hunk 3 above)
Comment (charlierudolph) [inline src/cli.coffee:14]: This was another debate I was having. Should requires used in a scope all be at the top or only required when they are used

File: src/cli.coffee (hunk 2 above)
Comment (charlierudolph) [inline src/cli.coffee]: Yes

File: src/cli.coffee (hunk 1 above)
Comment (charlierudolph) [inline src/cli.coffee]: Sorry snuck in another refactor here. This is unrelated. I'll revert

File: src/cli.coffee (hunk 4 above)
Comment (charlierudolph) [inline src/cli.coffee]: k

File: src/config/configuration_loader.coffee (hunk 5 above)
Comment (charlierudolph) [inline src/config/configuration_loader.coffee]: I'm fine pushing out anything required more than once into an outer scope to keep things dry.

I expect reviews to be able to catch your last point. And there's nothing stopping that from happening when all the requires were at the top of the file

Comment (kevgo) [inline src/config/configuration_loader.coffee]: Yeah. The only point is if all requires are at the top, its easier to see if another comparable library is already used. Having this buried somewhere on line 73 makes that harder. But this is not a strong point, and it only applies to messy code bases.

File: src/linter.coffee (hunk 6 above)
Comment (kevgo) [inline src/linter.coffee]: Re point 1: the complexity increases because the method now has an additional assignment and method call.

File: src/cli.coffee (hunk 3 above)
Comment (kevgo) [inline src/cli.coffee:14]: I was debating the same thing with myself  :)

Pros of moving requires up is that it makes code dry. Cons are that it makes refactoring harder, because once I use an external dependency a second time, I have to change the other method that uses it, even though this method isn't even related to my PR. 

Similarly, every time I remove a usage of an external dependency, I might have to move the require around to a more narrow scope. 

All of that sounds like a lot of unnecessary extra ceremony to me.

Comment (charlierudolph) [inline src/cli.coffee:14]: I'm thinking about a linter for this type of thing. One that checks that variable are declared in their minimal scope. I think it would be better in JSHint then coffeelint and thus I'm interested in trying out https://github.com/Clever/coffee-jshint

Comment (kevgo) [inline src/cli.coffee:14]: Cool idea!

**Comment by charlierudolph** [comment]
Like the idea and may use pieces but not using this entirely. Thanks for the feedback.
